File: packages/core/src/types/sources.ts
import { z } from "zod";

export type FamilyKind = "periodic" | "constant";
export type Granularity = "quarter" | "month" | "year";

/** Canonical period formats. Lexicographic order is chronological within one granularity. */
export const PERIOD_REGEX: Record<Granularity, RegExp> = {
  quarter: /^\d{4}-Q[1-4]$/,
  month: /^\d{4}-(0[1-9]|1[0-2])$/,
  year: /^\d{4}$/,
};

const MONTHS = ["Jan", "Feb", "Mar", "Apr", "May", "Jun", "Jul", "Aug", "Sep", "Oct", "Nov", "Dec"];

/** Display form of a canonical period; unknown shapes pass through untouched. */
export function formatPeriod(period: string): string {
  if (PERIOD_REGEX.quarter.test(period)) return `${period.slice(5)} ${period.slice(0, 4)}`;
  if (PERIOD_REGEX.month.test(period)) return `${MONTHS[Number(period.slice(5)) - 1]} ${period.slice(0, 4)}`;
  return period;
}

/** What the classifier proposes for one uploaded file (validated further in engine/classify). */
export const ClassifyProposalSchema = z.object({
  familyKey: z.string(),
  newFamily: z
    .object({
      key: z.string(),
      label: z.string(),
      kind: z.enum(["periodic", "constant"]),
      granularity: z.enum(["quarter", "month", "year"]).nullable(),
    })
    .optional(),
  period: z.string().nullable(),
  confidence: z.enum(["high", "medium", "low"]),
});
export type ClassifyProposal = z.infer<typeof ClassifyProposalSchema>;

export interface SourceFamilyRow {
  id: string;
  projectId: string;
  key: string;
  label: string;
  kind: FamilyKind;
  granularity: Granularity | null;
  createdAt: string;
}

export interface ProjectSourceRow {
  id: string;
  projectId: string;
  familyId: string | null;
  period: string | null;
  name: string;
  kind: string;
  storedFilename: string;
  mime: string;
  size: number;
  status: "unfiled" | "filed" | "replaced";
  proposal: ClassifyProposal | null;
  uploadedAt: string;
  filedAt: string | null;
}

File: backend/runoff_api/core/types/sources.py
"""Port of packages/core/src/types/sources.ts.

`PERIOD_REGEX` is a dict[str, re.Pattern] with the exact regex literals from the
TS source. `ClassifyProposalSchema` is a plain (non-strict) zod object -> the
CamelModelOpen `ClassifyProposal`. TS interface exports (`SourceFamilyRow`,
`ProjectSourceRow`) are compile-time-only and have no Python equivalent — those
rows flow as plain dicts at runtime.
"""

import re
from typing import Literal

from .base import CamelModelOpen
from .parse_plan import ExecReport, ParsePlan, PlanPreview

# Canonical period formats. Lexicographic order is chronological within one granularity.
PERIOD_REGEX: dict[str, re.Pattern] = {
    "quarter": re.compile(r"^\d{4}-Q[1-4]$"),
    "month": re.compile(r"^\d{4}-(0[1-9]|1[0-2])$"),
    "year": re.compile(r"^\d{4}$"),
}

MONTHS = ["Jan", "Feb", "Mar", "Apr", "May", "Jun", "Jul", "Aug", "Sep", "Oct", "Nov", "Dec"]


def format_period(period: str) -> str:
    """Port of formatPeriod. Display form of a canonical period; unknown shapes pass through."""
    if PERIOD_REGEX["quarter"].search(period):
        return f"{period[5:]} {period[0:4]}"
    if PERIOD_REGEX["month"].search(period):
        return f"{MONTHS[int(period[5:]) - 1]} {period[0:4]}"
    return period


class _NewFamily(CamelModelOpen):
    key: str
    label: str
    kind: Literal["periodic", "constant"]
    granularity: Literal["quarter", "month", "year"] | None


class _ClassifyTable(CamelModelOpen):
    name: str
    columns: list[str]
    row_count: float


class ClassifyProposal(CamelModelOpen):
    """What the classifier proposes for one uploaded file."""

    family_key: str
    new_family: _NewFamily | None = None
    period: str | None
    confidence: Literal["high", "medium", "low"]
    tables: list[_ClassifyTable] | None = None
    skipped_fragments: float | None = None
    drift: list[str] | None = None
    plan: ParsePlan | None = None
    plan_status: Literal["stored", "proposed", "amended", "none"] | None = None
    preview: PlanPreview | None = None
    report: ExecReport | None = None
